### PR TITLE
Allow player to delay start of friendly ground forces' advance on front line

### LINKED ISSUE #964: Allow player to delay start of friendly ground forces' advance on front line
**Is your feature request related to a problem? Please describe.**
Currently, friendly ground forces will often be decisively engaged (or dead) before CAS/BAI has a chance to get to the front line. This doesn't reflect reality where ground manoeuvre and supporting aircraft are closely synchronised in time.

**Describe the solution you'd like**
It would be good to allow the player to set a delay ground units starting their advance (utilise the same format as setting 'TOT' for aircraft packages). This would only need to apply to  '"aggressive, ""breakthrough" and "elimination" Frontline operations. "Defensive", "Retreat" and "Ambush" would not be effected by this and the enemy would conduct their operations per whatever timeline they have established. 
I have attached an image to show where this would be in the UI. The time should grey out if a Frontline Operation is selected that doesn't allow a delay to be implemented.

**Additional context**
To be clear, this is simply when the units *start* to move from initial positions, rather than arriving at a location (as per TOT).
delay timer

### PR DESCRIPTION
**Pull request by nosv1**
Implements https://github.com/dcs-liberation/dcs_liberation/issues/964

*eek... first gameplay PR*

So yeh, as the issue above suggests, an ability to delay the front line. As it is currently implemented, this does not consider the stance, but simply if the front line is delayed it is set to reach Waypoint 1 at the delayed time (Waypoint 1 is like a couple of feet in front of waypoint 0, so not really sure how the AI handles having WP 1 being a set time while still spawning in on mission start).

I think that's it really... I never did test an actual mission, but the ME showed everything as expected.

The delayed start should be getting reset to 0 after each turn.

image
image
image

it's the ground speed of 0 that has me worried for WP 1

### CODE REVIEW

**Comment by bgreman** [comment]
Reopening after inadvertent branch deletion

File: game/theater/controlpoint.py [hunk 1]
```diff
@@ -309,6 +310,7 @@ def __init__(
         self.captured_invert = False
         # TODO: Should be Airbase specific.
         self.has_frontline = has_frontline
+        self.delayed_start: timedelta = timedelta(seconds=0)
```
Comment (DanAlbert) [inline game/theater/controlpoint.py:313]: It looks like you're creating a `QTimeEdit` per front but only a delay per base. The most recently set time will clobber the others for a multi-front base.

File: gen/armor.py [hunk 2]
```diff
@@ -545,6 +546,14 @@ def plan_action_for_groups(
                 dcs_group.add_waypoint(retreat_point, PointAction.OffRoad)
                 dcs_group.add_waypoint(reposition_point, PointAction.OffRoad)
 
+            if len(dcs_group.points) >= 2:
+                try:
+                    dcs_group.points[1].ETA = from_cp.delayed_start.total_seconds()
+                    dcs_group.points[1].ETA_locked = True
+                    dcs_group.points[1].speed_locked = False
+                except AttributeError:  # old saves may not have delayed_start
+                    pass
```
Comment (DanAlbert) [inline gen/armor.py:555]: A better way to do this is with `__setstate__`. There are examples of how to do it elsewhere in the codebase.

File: gen/armor.py [hunk 3]
```diff
@@ -545,6 +546,14 @@ def plan_action_for_groups(
                 dcs_group.add_waypoint(retreat_point, PointAction.OffRoad)
                 dcs_group.add_waypoint(reposition_point, PointAction.OffRoad)
 
+            if len(dcs_group.points) >= 2:
+                try:
+                    dcs_group.points[1].ETA = from_cp.delayed_start.total_seconds()
+                    dcs_group.points[1].ETA_locked = True
+                    dcs_group.points[1].speed_locked = False
```
Comment (DanAlbert) [inline gen/armor.py:553]: The fact that WP1 _must_ be close to WP0 for this to work is a really easy thing to overlook for anyone that might be altering front line behavior. Instead, create a wait action with an end time like we do for hold points for air units?

File: qt_ui/windows/basemenu/ground_forces/QGroundForcesStrategy.py [hunk 4]
```diff
@@ -31,6 +33,18 @@ def cheat() -> None:
...
                 layout.addWidget(QLabel(enemy_cp.name))
                 layout.addWidget(QGroundForcesStrategySelector(self.cp, enemy_cp))
+
+                layout.addWidget(QLabel("Delay movement until"))
+                self.delayed_start_spinner = QTimeEdit()
+                self.delayed_start_spinner.setMinimumTime(QTime(0, 0))
+                self.delayed_start_spinner.setDisplayFormat("T+hh:mm:ss")
+                self.delayed_start_spinner.timeChanged.connect(self.save_delayed_start)
+                self.delayed_start_spinner.setToolTip(
+                    "Forces will not move from Waypoint 1 until the specified time has elapsed."
```
Comment (DanAlbert) [inline qt_ui/windows/basemenu/ground_forces/QGroundForcesStrategy.py:43]: Players don't know what waypoint 1 is (I didn't until I saw the screenshot in this PR).

File: qt_ui/windows/basemenu/ground_forces/QGroundForcesStrategy.py [hunk 5]
```diff
@@ -59,3 +73,8 @@ def cheat_alter_front_line(self, enemy_point: ControlPoint, advance: bool) -> No
         self.game.reset_ato()
         self.game.initialize_turn()
         GameUpdateSignal.get_instance().updateGame(self.game)
+
+    def save_delayed_start(self) -> None:
+        time = self.delayed_start_spinner.time()
+        seconds = time.hour() * 3600 + time.minute() * 60 + time.second()
+        self.cp.delayed_start = timedelta(seconds=seconds)
```
Comment (DanAlbert) [inline qt_ui/windows/basemenu/ground_forces/QGroundForcesStrategy.py:80]:
```suggestion
        self.cp.delayed_start = timedelta(hours=time.hour(), minutes=time.minute(), seconds=time.second())
```

**Review by DanAlbert — CHANGES_REQUESTED** [review]

**Comment by DanAlbert** [comment]
Closing stale PRs for clarity. They're trivial to reopen if you want to return to this :)

**Pull request closed without merging**